File: Guigna/Guigna/GItem.m
#import "GItem.h"

#import "GAdditions.h"

@implementation GItem

@synthesize description;

- (instancetype)initWithName:(NSString *)name
           version:(NSString *)version
            source:(GSource *)source
            status:(GStatus)status {
    self = [super init];
    if (self) {
        _name = [name copy];
        _version = [version copy];
        _source = source;
        _status = status;
    }
    return self;
}

// TODO:

- (instancetype)initWithCoder:(NSCoder *)coder
{
	self = [super init];
	if (self) {
		_name = [coder decodeObjectForKey:@"name"];
		_version = [coder decodeObjectForKey:@"version"];
	}
	return self;
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:_name forKey:@"name"];
    [coder encodeObject:_version forKey:@"version"];
}


- (NSString *)info {
    return [self.source info:self];
}

- (NSString *)home {
    return [self.source home:self];
}

- (NSString *)log {
    return [self.source log:self];
}

- (NSString *)contents {
    return [self.source contents:self];
}

- (NSString *)cat {
    return [self.source cat:self];
}

- (NSString *)deps {
    return [self.source deps:self];
}

- (NSString *)dependents {
    return [self.source dependents:self];
}


@end


@implementation GStatusTransformer

+ (Class)transformedValueClass {
	return [NSImage class];
}

+ (BOOL)allowsReverseTransformation {
	return NO;
}

- (id)transformedValue:(id)status {
    switch ([status intValue]) {
        case GInactiveStatus:
            return [NSImage imageNamed:NSImageNameStatusNone];
            break;
        case GUpToDateStatus:
            return [NSImage imageNamed:NSImageNameStatusAvailable];
            break;
        case GOutdatedStatus:
            return [NSImage imageNamed:NSImageNameStatusPartiallyAvailable];
            break;
        case GUpdatedStatus:
            return [NSImage imageNamed:@"status-updated.tiff"];
            break;
        case GNewStatus:
            return [NSImage imageNamed:@"status-new.tiff"];
            break;
        case GBrokenStatus:
            return [NSImage imageNamed:NSImageNameStatusUnavailable];
            break;
        default:
            return nil;
    }
}

@end


@implementation GMarkTransformer

+ (Class)transformedValueClass {
	return [NSImage class];
}

+ (BOOL)allowsReverseTransformation {
	return NO;
}

- (id)transformedValue:(id)mark {
    switch ([mark intValue]) {
        case GInstallMark:
            return [NSImage imageNamed:NSImageNameAddTemplate];
            break;
        case GUninstallMark:
            return [NSImage imageNamed:NSImageNameRemoveTemplate];
            break;
        case GDeactivateMark:
            return [NSImage imageNamed:NSImageNameStopProgressTemplate];
            break;
        case GUpgradeMark:
            return [NSImage imageNamed:NSImageNameRefreshTemplate];
            break;
        case GFetchMark:
            return [NSImage imageNamed:@"source-native.tiff"];
            break;
        case GCleanMark:
            return [NSImage imageNamed:NSImageNameActionTemplate];
            break;
        default:
            return nil;
    }
}

@end

File: Guigna-Swift/Guigna/GItem.swift
import Foundation

enum GStatus: Int {
    case Available = 0
    case Inactive
    case UpToDate
    case Outdated
    case Updated
    case New
    case Broken
}

enum GMark: Int {
    case NoMark = 0
    case Install
    case Uninstall
    case Deactivate
    case Upgrade
    case Fetch
    case Clean
}


class GItem: NSObject {
    var name: String
    var version: String
    weak var source: GSource!
    weak var system: GSystem!
    
    var status: GStatus {
    willSet(newValue) {
        self.willChangeValueForKey("statusValue")
    }
    didSet(newValue) {
        self.didChangeValueForKey("statusValue")
    }
    }
    var statusValue: NSNumber {
    get {
        return NSNumber.numberWithInteger(self.status.toRaw())
    }
    set {
        status = GStatus.fromRaw(newValue.integerValue)!
    }
    }
    
    var mark: GMark {
    willSet(newValue) {
        self.willChangeValueForKey("markValue")
    }
    didSet(newValue) {
        self.didChangeValueForKey("markValue")
    }
    }
    var markValue: NSNumber {
    get {
        return NSNumber.numberWithInteger(self.mark.toRaw())
    }
    set {
        mark = GMark.fromRaw(newValue.integerValue)!
    }
    }
    
    var installed: String!
    var categories: String?
    var _description: String?
    override var description: String {
    get {
        if _description {
            return _description!
        } else {
            return ""
        }
    }
    set {
        self._description = newValue}
    }
    var homepage: String!
    var screenshots: String!
    var URL: String!
    var license: String?
    var id: String!
    
    init(name: String, version: String, source: GSource, status: GStatus) {
        self.name = name
        self.version = version
        self.source = source
        self.system = nil
        self.status = status
        self.mark = .NoMark
    }
    
    var info: String {
    get {
        return source.info(self)
    }
    }
    
    var home: String {
    get {
        return source.home(self)
    }
    }
    
    var log: String {
    get {
        return source.log(self)
    }
    }
    
    var contents: String {
    get {
        return source.contents(self)
    }
    }
    
    var cat: String {
    get {
        return source.cat(self)
    }
    }
    
    var deps: String {
    get {
        return source.deps(self)
    }
    }
    
    var dependents: String {
    get {
        return source.dependents(self)
    }
    }
}


@objc(GStatusTransformer)
class GStatusTransformer: NSValueTransformer {
    
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    override func transformedValue(value: AnyObject!) -> AnyObject! {
        if !value {
            return nil
        }
        let status = GStatus.fromRaw(value.integerValue)!
        switch status {
        case .Inactive:
            return NSImage(named: NSImageNameStatusNone)
        case .UpToDate:
            return NSImage(named: NSImageNameStatusAvailable)
        case .Outdated:
            return NSImage(named: NSImageNameStatusPartiallyAvailable)
        case .Updated:
            return NSImage(named: "status-updated.tiff")
        case .New:
            return NSImage(named: "status-new.tiff")
        case .Broken:
            return NSImage(named: NSImageNameStatusUnavailable)
        default:
            return nil
        }
    }
}


@objc(GMarkTransformer)
class GMarkTransformer: NSValueTransformer {
    
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    override func transformedValue(value: AnyObject!) -> AnyObject! {
        if !value  {
            return nil
        }
        let mark = GMark.fromRaw(value.integerValue)!
        switch mark {
        case .Install:
            return NSImage(named: NSImageNameAddTemplate)
        case .Uninstall:
            return NSImage(named: NSImageNameRemoveTemplate)
        case .Deactivate:
            return NSImage(named: NSImageNameStopProgressTemplate)
        case .Upgrade:
            return NSImage(named: NSImageNameRefreshTemplate)
        case .Fetch:
            return NSImage(named: "source-native.tiff")
        case .Clean:
            return NSImage(named: NSImageNameActionTemplate)
        default:
            return nil
        }
    }
}


